src/collector/mod.rs

```diff
@@ -0,0 +1,210 @@
+//!   such cases.
+//!   e.g. attaching to kfree_skb_reason in the Linux kernel is better than
+//!   attaching to kfree_skb, as the drop reason is otherwise lost, but it is
+//!   acceptable as a fallback (mainly for older kernels).
+
+use std::collections::HashMap;
+use anyhow::{bail, Result};
+use log::{error, warn};
+
+/// Generic trait representing a collector. All collectors are required to
+/// implement this, as they'll be manipulated through this trait.
+trait Collector {
+    /// Allocate and return a new instance of the collector, using only default
+    /// values for its internal fields.
+    fn new() -> Result<Self> where Self: Sized;
+    /// Return the name of the collector. It *has* to be unique among all the
+    /// collectors.
+    fn name(&self) -> &'static str;
+    /// Initialize the collector, likely to be used to pass configuration data
+    /// such as filters or command line arguments. We need to split the new &
+    /// the init phase for collectors, to allow giving information to the core
+    /// as part of the collector registration and only then feed the collector
+    /// with data coming from the core. Checks for the mandatory part of the
+    /// collector should be done here.
+    fn init(&mut self) -> Result<()>;
+    /// Start the collection of events.
+    fn start(&mut self) -> Result<()>;
+}
+
+/// Collections of collectors. Used to handle a set of collectors and to perform
+/// group actions.
+pub(crate) struct CollectorsCollection {
+    list: HashMap<String, Box<dyn Collector>>,
+}
+
+impl CollectorsCollection {
+    fn new() -> CollectorsCollection {
+        CollectorsCollection {
+            list: HashMap::new(),
+        }
+    }
+
+    /// Register a collector to the collector collection.
+    ///
+    /// ```
+    /// collection
+    ///     .register(Box::new(FirstCollector::new()?))?
+    ///     .register(Box::new(SecondCollector::new()?))?
+    ///     .register(Box::new(ThirdCollector::new()?))?;
+    /// ```
+    fn register(&mut self, collector: Box<dyn Collector>) -> Result<&mut Self> {
+        let name = String::from(collector.name());
+
+        // Ensure uniqueness of the collector name. This is important as their
+        // name is used as a key.
+        if let Some(_) = self.list.get(&name) {
+            bail!("Could not insert collector '{}'; name already registered", name);
+        }
+
+        self.list.insert(name, collector);
+        Ok(self)
+    }
+
+    /// Initialize all collectors by calling their `init()` function. Collectors
+    /// failing to initialize will be removed from the collection.
+    pub(crate) fn init(&mut self) -> Result<()> {
+        let mut to_remove = Vec::new();
+
+        // Try initializing all collectors in the collection. Failing ones are
+        // put on a list for future removal.
+        for (_, c) in self.list.iter_mut() {
+            if let Err(_) = c.init() {
+                to_remove.push(c.name());
```

Comment:
Maybe move the above `error!` up here to log the initialization error alongside what c.init() returns?